index.ts

```diff
@@ -87,28 +79,10 @@ export const decode = (input: string) => {
  * @private
  * @param {string} string - Text to encode
  */
-function _utf8_encode(string: string) {
-  string = string.replace(/\r\n/g, '\n');
-  let utftext = '';
-
-  for (let n = 0; n < string.length; n++) {
-    const c = string.charCodeAt(n);
-
-    if (c < 128) {
-      utftext += String.fromCharCode(c);
-    } else if (c > 127 && c < 2048) {
-      utftext +=
-        String.fromCharCode((c >> 6) | 192) +
-        String.fromCharCode((c & 63) | 128);
-    } else {
-      utftext +=
-        String.fromCharCode((c >> 12) | 224) +
-        String.fromCharCode(((c >> 6) & 63) | 128) +
-        String.fromCharCode((c & 63) | 128);
-    }
-  }
-
-  return utftext;
+function _utf8_encode(str: string): string {
+  return encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, (_, p1) =>
```

Comment:
Hm, I'm not sure about using the `encodeURIComponent` method - it had some gimmicks which blocked me from using it in the first place.
Pity I didn't write tests for that, give me few days to come up with reproduction of the issue I'm talking about - if I don't find it, we can go with your change

Reply:
It will finally get the same result as TextEncoder, and if acceptable, use TextEncoder and TextDecoder can lead to a smaller bundle size.
For better compatibility I chose the `encodeURIComponent`, it will also use utf-8 inside to encode string, the 4 bytes unicode character will be converted into something like `%aa%bb%cc%dd`, when we converted it into 4 separated character, it will be the same thing as what TextEncoder.encode does at bytes level.

Reply:
Okay, so:

This one case that I was talking about - it was about ability to base64 encode/decode strings that are outside of standard Latin1 character range (e.g. special characters with diacritics like ąęś or just non-latin characters like 码). But your implementation handles those correctly as you're not using `atob` or `btoa` functions that are known for not handling special characters well 🎉 

I've added unit tests on the main branch, so please rebase your branch to get them. I've already prepared for you a test case [covering the issue you've reported](https://github.com/FRSOURCE/base64/blob/main/index.spec.ts#L27) (it's commented-out for now, so please just uncomment it).

Once the tests will run successfully on your branch I will happily merge this PR ❤️

Reply:
> Okay, so:
> 
> This one case that I was talking about - it was about ability to base64 encode/decode strings that are outside of standard Latin1 character range (e.g. special characters with diacritics like ąęś or just non-latin characters like 码). But your implementation handles those correctly as you're not using `atob` or `btoa` functions that are known for not handling special characters well 🎉
> 
> I've added unit tests on the main branch, so please rebase your branch to get them. I've already prepared for you a test case [covering the issue you've reported](https://github.com/FRSOURCE/base64/blob/main/index.spec.ts#L27) (it's commented-out for now, so please just uncomment it).
> 
> Once the tests will run successfully on your branch I will happily merge this PR ❤️

I just rebased my branch and enable all the tests, they were all passed, you can check it later when you're available ❤